brain/detail/circuit.h

```diff
@@ -222,12 +189,335 @@ class Circuit::Impl
+    virtual void saveMorphologyToCache(
+        const std::string& uri, const std::string& hash,
+        neuron::MorphologyPtr morphology) const = 0;
+
+    virtual CachedMorphologies loadMorphologiesFromCache(
+        const std::set<std::string>& hashes) const = 0;
+    virtual void saveSynapsePositionsToCache(
+        const uint32_t gid, const std::string& hash,
+        const brion::SynapseMatrix& value) const = 0;
+    virtual CachedSynapses loadSynapsePositionsFromCache(
+        const Strings& keys) const = 0;
+    virtual void _findSynapsePositionsColumns() const = 0;
+    virtual const brion::SynapseSummary& getSynapseSummary() const = 0;
+    virtual const brion::Synapse& getSynapseAttributes(
+        const bool afferent) const = 0;
+    virtual const brion::Synapse& getAfferentProjectionAttributes(
+        const std::string& name) const = 0;
+    virtual const brion::Synapse* getSynapseExtra() const = 0;
+    virtual const brion::Synapse& getSynapsePositions(
+        const bool afferent) const = 0;
+    virtual const URI& getSynapseSource() const = 0;
+};
+
+class SonataCircuit : public Circuit::Impl
+{
+public:
+    explicit SonataCircuit(const URI& uri)
+        : config(uri)
+        , basePath(boost::filesystem::path(uri.getPath()).parent_path())
+    {
+        const auto& subnetworks = config.getNodes();
+        const auto& node = subnetworks.front();
+        brion::Nodes nodes(URI(basePath.string() + "/" + node.elements));
+
+        if (subnetworks.size() > 1)
+        {
+            LBWARN << "More than one subnetwork found, ignored." << std::endl;
+        }
+
+        const auto populations = nodes.getPopulationNames();
+        if (populations.size() > 1)
+        {
+            LBWARN << "More than one population found, ignored." << std::endl;
+        }
+
+        const auto population = populations.front();
+        const auto nodeGroupIDs = nodes.getNodeGroupIDs(population);
+        const auto nodeIDs = nodes.getNodeIDs(population);
+        const auto nodeGroupIndices = nodes.getNodeGroupIndices(population);
+        const size_t numNodes = nodeIDs.size();
+
+        if (nodeGroupIDs.size() > 1)
+        {
+            LBWARN << "More than one group ID found, ignored." << std::endl;
+        }
+
+        nodeGroup = nodes.openGroup(population, 0);
+
+        size_t expectedIdx = 0;
+        for (size_t i = 0; i < numNodes; i++)
+        {
+            const auto groupID = nodeGroupIDs[i];
+            const auto nodeGroupIndex = nodeGroupIndices[i];
+
+            if (groupID != 0)
+                continue;
+
+            if (expectedIdx != nodeGroupIndex)
+            {
+                LBTHROW(
+                    std::runtime_error("Expected node group index '" +
+                                       std::to_string(expectedIdx) + "' got '" +
+                                       std::to_string(nodeGroupIndex) + "'"));
+            }
+
+            expectedIdx++;
+        }
```

Comment:
I would just compare nodeGroupIndices to a vector created with std::iota, it's shorter and it doesn't really matter which index is out of order because later we have to implement the indirect code.

Reply:
I do not understand how to do that so I will leave like this since we are removing it anyway